// kuvork/dist/vehicles/european.vehicle.hyundai.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { REGIONS, DEFAULT_VEHICLE_STATUS_OPTIONS } from '../constants';
import got from 'got';
import logger from '../logger';
import { Vehicle } from './vehicle';
import { getTempCode } from '../util';
import { EU_BASE_URL } from '../constants/europe';
var EuropeanVehicle = /** @class */ (function (_super) {
    __extends(EuropeanVehicle, _super);
    function EuropeanVehicle(vehicleConfig, controller) {
        var _this = _super.call(this, vehicleConfig, controller) || this;
        _this.vehicleConfig = vehicleConfig;
        _this.controller = controller;
        _this.region = REGIONS.EU;
        logger.debug("EU Vehicle " + _this.vehicleConfig.id + " created");
        return _this;
    }
    EuropeanVehicle.prototype.checkControlToken = function () {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!(((_a = this.controller.session) === null || _a === void 0 ? void 0 : _a.controlTokenExpiresAt) !== undefined)) return [3 /*break*/, 2];
                        if (!(this.controller.session.controlToken === '' ||
                            new Date().getTime() > this.controller.session.controlTokenExpiresAt)) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.controller.enterPin()];
                    case 1:
                        _b.sent();
                        _b.label = 2;
                    case 2: return [2 /*return*/];
                }
            });
        });
    };
    EuropeanVehicle.prototype.start = function (config) {
        return __awaiter(this, void 0, void 0, function () {
            var response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.checkControlToken()];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, got(EU_BASE_URL + "/api/v2/spa/vehicles/" + this.vehicleConfig.id + "/control/temperature", {
                                method: 'POST',
                                body: {
                                    action: 'start',
                                    hvacType: 0,
                                    options: {
                                        defrost: config.defrost,
                                        heating1: config.windscreenHeating ? 1 : 0,
                                    },
                                    tempCode: getTempCode(config.temperature),
                                    unit: config.unit,
                                },
                                headers: {
                                    'Authorization': this.controller.session.controlToken,
                                    'ccsp-device-id': this.controller.session.deviceId,
                                    'Content-Type': 'application/json',
                                },
                                json: true,
                            })];
                    case 2:
                        response = _a.sent();
                        logger.info("Climate started for vehicle " + this.vehicleConfig.id);
                        return [2 /*return*/, Promise.resolve(response.body)];
                }
            });
        });
    };
    EuropeanVehicle.prototype.stop = function () {
        return __awaiter(this, void 0, void 0, function () {
            var response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.checkControlToken()];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, got(EU_BASE_URL + "/api/v2/spa/vehicles/" + this.vehicleConfig.id + "/control/temperature", {
                                method: 'POST',
                                body: {
                                    action: 'stop',
                                    hvacType: 0,
                                    options: {
                                        defrost: true,
                                        heating1: 1,
                                    },
                                    tempCode: '10H',
                                    unit: 'C',
                                },
                                headers: {
                                    'Authorization': this.controller.session.controlToken,
                                    'ccsp-device-id': this.controller.session.deviceId,
                                    'Content-Type': 'application/json',
                                },
                                json: true,
                            })];
                    case 2:
                        response = _a.sent();
                        logger.info("Climate stopped for vehicle " + this.vehicleConfig.id);
                        return [2 /*return*/, Promise.resolve(response.body)];
                }
            });
        });
    };
    EuropeanVehicle.prototype.lock = function () {
        return __awaiter(this, void 0, void 0, function () {
            var response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.checkControlToken()];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, got(EU_BASE_URL + "/api/v2/spa/vehicles/" + this.vehicleConfig.id + "/control/door", {
                                method: 'POST',
                                headers: {
                                    'Authorization': this.controller.session.controlToken,
                                    'ccsp-device-id': this.controller.session.deviceId,
                                    'Content-Type': 'application/json',
                                },
                                body: {
                                    action: 'close',
                                    deviceId: this.controller.session.deviceId,
                                },
                                json: true,
                            })];
                    case 2:
                        response = _a.sent();
                        if (response.statusCode === 200) {
                            logger.debug("Vehicle " + this.vehicleConfig.id + " locked");
                            return [2 /*return*/, Promise.resolve('Lock successful')];
                        }
                        return [2 /*return*/, Promise.reject('Something went wrong!')];
                }
            });
        });
    };
    EuropeanVehicle.prototype.unlock = function () {
        return __awaiter(this, void 0, void 0, function () {
            var response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.checkControlToken()];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, got(EU_BASE_URL + "/api/v2/spa/vehicles/" + this.vehicleConfig.id + "/control/door", {
                                method: 'POST',
                                headers: {
                                    'Authorization': this.controller.session.controlToken,
                                    'ccsp-device-id': this.controller.session.deviceId,
                                    'Content-Type': 'application/json',
                                },
                                body: {
                                    action: 'open',
                                    deviceId: this.controller.session.deviceId,
                                },
                                json: true,
                            })];
                    case 2:
                        response = _a.sent();
                        if (response.statusCode === 200) {
                            logger.debug("Vehicle " + this.vehicleConfig.id + " unlocked");
                            return [2 /*return*/, Promise.resolve('Unlock successful')];
                        }
                        return [2 /*return*/, Promise.reject('Something went wrong!')];
                }
            });
        });
    };
    EuropeanVehicle.prototype.status = function (input) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function () {
            var statusConfig, cacheString, response, vehicleStatus, parsedStatus;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        statusConfig = __assign(__assign({}, DEFAULT_VEHICLE_STATUS_OPTIONS), input);
                        return [4 /*yield*/, this.checkControlToken()];
                    case 1:
                        _c.sent();
                        cacheString = statusConfig.refresh ? '' : '/latest';
                        return [4 /*yield*/, got(EU_BASE_URL + "/api/v2/spa/vehicles/" + this.vehicleConfig.id + "/status" + cacheString, {
                                method: 'GET',
                                headers: {
                                    'Authorization': this.controller.session.controlToken,
                                    'ccsp-device-id': this.controller.session.deviceId,
                                    'Content-Type': 'application/json',
                                },
                                json: true,
                            })];
                    case 2:
                        response = _c.sent();
                        if (statusConfig.refresh)
                            vehicleStatus = response.body.resMsg;
                        else
                            vehicleStatus = response.body.resMsg.vehicleStatusInfo.vehicleStatus;
                        parsedStatus = {
                            chassis: {
                                hoodOpen: vehicleStatus.hoodOpen,
                                trunkOpen: vehicleStatus.trunkOpen,
                                locked: vehicleStatus.doorLock,
                                doors: {
                                    frontRight: !!vehicleStatus.doorOpen.frontRight,
                                    frontLeft: !!vehicleStatus.doorOpen.frontLeft,
                                    backLeft: !!vehicleStatus.doorOpen.backLeft,
                                    backRight: !!vehicleStatus.doorOpen.backRight,
                                },
                                tirePressureWarningLamp: {
                                    rearLeft: !!vehicleStatus.tirePressureLamp.tirePressureLampRL,
                                    frontLeft: !!vehicleStatus.tirePressureLamp.tirePressureLampFL,
                                    frontRight: !!vehicleStatus.tirePressureLamp.tirePressureLampFR,
                                    rearRight: !!vehicleStatus.tirePressureLamp.tirePressureLampRR,
                                    all: !!vehicleStatus.tirePressureLamp.tirePressureLampAll,
                                },
                            },
                            climate: {
                                active: vehicleStatus.airCtrlOn,
                                steeringwheelHeat: !!vehicleStatus.steerWheelHeat,
                                sideMirrorHeat: false,
                                rearWindowHeat: !!vehicleStatus.sideBackWindowHeat,
                                defrost: vehicleStatus.defrost,
                                temperatureSetpoint: vehicleStatus.airTemp.value,
                                temperatureUnit: vehicleStatus.airTemp.unit,
                            },
                            engine: {
                                ignition: vehicleStatus.engine,
                                adaptiveCruiseControl: vehicleStatus.acc,
                                range: vehicleStatus.evStatus.drvDistance[0].rangeByFuel.totalAvailableRange.value,
                                charging: (_a = vehicleStatus === null || vehicleStatus === void 0 ? void 0 : vehicleStatus.evStatus) === null || _a === void 0 ? void 0 : _a.batteryCharge,
                                batteryCharge: (_b = vehicleStatus === null || vehicleStatus === void 0 ? void 0 : vehicleStatus.battery) === null || _b === void 0 ? void 0 : _b.batSoc,
                            }
                        };
                        this._status = input.parsed ? parsedStatus : vehicleStatus;
                        return [2 /*return*/, Promise.resolve(this._status)];
                }
            });
        });
    };
    EuropeanVehicle.prototype.odometer = function () {
        return __awaiter(this, void 0, void 0, function () {
            var response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.checkControlToken()];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, got(EU_BASE_URL + "/api/v2/spa/vehicles/" + this.vehicleConfig.id + "/status/latest", {
                                method: 'GET',
                                headers: {
                                    'Authorization': this.controller.session.controlToken,
                                    'ccsp-device-id': this.controller.session.deviceId,
                                    'Content-Type': 'application/json',
                                },
                                json: true,
                            })];
                    case 2:
                        response = _a.sent();
                        this._odometer = response.body.resMsg.vehicleStatusInfo.odometer;
                        return [2 /*return*/, Promise.resolve(this._odometer)];
                }
            });
        });
    };
    EuropeanVehicle.prototype.location = function () {
        return __awaiter(this, void 0, void 0, function () {
            var response, data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.checkControlToken()];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, got(EU_BASE_URL + "/api/v2/spa/vehicles/" + this.vehicleConfig.id + "/location", {
                                method: 'GET',
                                headers: {
                                    'Authorization': this.controller.session.controlToken,
                                    'ccsp-device-id': this.controller.session.deviceId,
                                    'Content-Type': 'application/json',
                                },
                                json: true,
                            })];
                    case 2:
                        response = _a.sent();
                        data = response.body.resMsg.gpsDetail;
                        this._location = {
                            latitude: data.coord.lat,
                            longitude: data.coord.lon,
                            altitude: data.coord.alt,
                            speed: {
                                unit: data.speed.unit,
                                value: data.speed.value,
                            },
                            heading: data.head,
                        };
                        return [2 /*return*/, Promise.resolve(this._location)];
                }
            });
        });
    };
    return EuropeanVehicle;
}(Vehicle));
export default EuropeanVehicle;
