// kuvork/dist/vehicles/vehicle.js
import { REGIONS } from '../constants';
var Vehicle = /** @class */ (function () {
    function Vehicle(vehicleConfig, controller) {
        this.vehicleConfig = vehicleConfig;
        this.controller = controller;
        this._status = null;
        this._location = null;
        this._odometer = null;
        this.userConfig = {
            username: undefined,
            password: undefined,
            region: REGIONS.EU,
            autoLogin: true,
            pin: undefined,
            vin: undefined,
            vehicleId: undefined,
        };
        this.userConfig = controller.userConfig;
    }
    Vehicle.prototype.vin = function () {
        return this.vehicleConfig.vin;
    };
    Vehicle.prototype.name = function () {
        return this.vehicleConfig.name;
    };
    Vehicle.prototype.nickname = function () {
        return this.vehicleConfig.nickname;
    };
    Vehicle.prototype.id = function () {
        return this.vehicleConfig.id;
    };
    Vehicle.prototype.brandIndicator = function () {
        return this.vehicleConfig.brandIndicator;
    };
    return Vehicle;
}());
export { Vehicle };
